### PR TITLE
[autoscaling] increase connect timeout, boto retries, and check subnet conf

### PR DESCRIPTION
**Pull request by ericl**
Some quick fixes to help mitigate https://github.com/ray-project/ray/issues/1418 and https://github.com/ray-project/ray/issues/1420

### CODE REVIEW

**Comment by robertnishihara** [comment]
Just tried running this (on the PR branch), and got

```
Role not specified for head node, using arn:aws:iam::339530224232:instance-profile/ray-autoscaler
KeyName not specified for nodes, using ray-autoscaler_1
Traceback (most recent call last):
  File "/Users/rkn/anaconda3/bin/ray", line 11, in <module>
    load_entry_point('ray', 'console_scripts', 'ray')()
  File "/Users/rkn/Workspace/ray/python/ray/scripts/scripts.py", line 276, in main
    return cli()
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/rkn/Workspace/ray/python/ray/scripts/scripts.py", line 260, in create_or_update
    cluster_config_file, min_workers, max_workers, no_restart)
  File "/Users/rkn/Workspace/ray/python/ray/autoscaler/commands.py", line 39, in create_or_update_cluster
    config = bootstrap_config(config)
  File "/Users/rkn/Workspace/ray/python/ray/autoscaler/aws/config.py", line 40, in bootstrap_aws
    config = _configure_subnet(config)
  File "/Users/rkn/Workspace/ray/python/ray/autoscaler/aws/config.py", line 163, in _configure_subnet
    "The chosen subnet must map nodes with public IPs on launch"
AssertionError: The chosen subnet must map nodes with public IPs on launch
```

File: python/ray/autoscaler/aws/config.py [hunk 1]
```diff
@@ -8,20 +8,25 @@
...
     """Returns the ith default (aws_key_pair_name, key_pair_path)."""
     if i == 0:
-        return RAY, os.path.expanduser("~/.ssh/{}.pem".format(RAY))
+        return (
+            "{}_{}".format(RAY, region),
+            os.path.expanduser("~/.ssh/{}_{}.pem".format(RAY, region)))
     return (
-        "{}_{}".format(RAY, i),
-        os.path.expanduser("~/.ssh/{}_{}.pem".format(RAY, i)))
+        "{}_{}_{}".format(RAY, i, region),
```
Comment (ericl) [inline python/ray/autoscaler/aws/config.py]: turns out key pairs are region local

**Comment by ericl** [comment]
I added a filter to avoid those subnets, which should fix that I think. Also tweaked the boto config to increase the max number of retries (otherwise you sometimes get rate limit exceeded when running with very large clusters)

**Comment by robertnishihara** [comment]
Just tried again and got

```
$ ray create_or_update ~/Workspace/ray/python/ray/autoscaler/aws/example.yaml
Traceback (most recent call last):
  File "/Users/rkn/anaconda3/bin/ray", line 11, in <module>
    load_entry_point('ray', 'console_scripts', 'ray')()
  File "/Users/rkn/Workspace/ray/python/ray/scripts/scripts.py", line 276, in main
    return cli()
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/rkn/Workspace/ray/python/ray/scripts/scripts.py", line 260, in create_or_update
    cluster_config_file, min_workers, max_workers, no_restart)
  File "/Users/rkn/Workspace/ray/python/ray/autoscaler/commands.py", line 39, in create_or_update_cluster
    config = bootstrap_config(config)
  File "/Users/rkn/Workspace/ray/python/ray/autoscaler/aws/config.py", line 39, in bootstrap_aws
    config = _configure_iam_role(config)
  File "/Users/rkn/Workspace/ray/python/ray/autoscaler/aws/config.py", line 58, in _configure_iam_role
    profile = _get_instance_profile(DEFAULT_RAY_INSTANCE_PROFILE, config)
  File "/Users/rkn/Workspace/ray/python/ray/autoscaler/aws/config.py", line 256, in _get_instance_profile
    iam = _resource("iam", config)
  File "/Users/rkn/Workspace/ray/python/ray/autoscaler/aws/config.py", line 279, in _resource
    boto_config = Config(retries=dict(max_attempts=BOTO_MAX_RETRIES))
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/botocore-1.5.45-py3.6.egg/botocore/config.py", line 100, in __init__
    args, kwargs)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/botocore-1.5.45-py3.6.egg/botocore/config.py", line 125, in _record_user_provided_options
    'Got unexpected keyword argument \'%s\'' % key)
TypeError: Got unexpected keyword argument 'retries'
```

Note 

```python
import botocore
import boto3

print(botocore.__version__)  # '1.5.45'
print(boto3.__version__)  # '1.4.4'
```

**Comment by ericl** [comment]
Hm I'm going to add a check for boto version >= 1.4.8, the older ones don't support this config or spot instances.

**Comment by robertnishihara** [comment]
Note I still run into #1407 when trying `development-example.yaml`. We should fix that (perhaps in a separate PR) but not by adding in the availability zone to the yaml file.

Upgrading boto3 worked. However, the first time I tried to upgrade boto3 with `pip install -U boto3`, I got the error

```
Successfully installed boto3-1.5.15 botocore-1.8.29 docutils-0.14 jmespath-0.9.3 s3transfer-0.1.12
Traceback (most recent call last):
  File "/Users/rkn/anaconda3/bin/pip", line 6, in <module>
    sys.exit(pip.main())
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/__init__.py", line 249, in main
    return command.main(cmd_args)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/basecommand.py", line 252, in main
    pip_version_check(session)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/utils/outdated.py", line 102, in pip_version_check
    installed_version = get_installed_version("pip")
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/utils/__init__.py", line 838, in get_installed_version
    working_set = pkg_resources.WorkingSet()
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 644, in __init__
    self.add_entry(entry)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 700, in add_entry
    for dist in find_distributions(entry, True):
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 1949, in find_eggs_in_zip
    if metadata.has_metadata('PKG-INFO'):
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 1463, in has_metadata
    return self.egg_info and self._has(self._fn(self.egg_info, name))
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 1823, in _has
    return zip_path in self.zipinfo or zip_path in self._index()
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 1703, in zipinfo
    return self._zip_manifests.load(self.loader.archive)
  File "/Users/rkn/anaconda3/lib/python3.6/site-packages/pip/_vendor/pkg_resources/__init__.py", line 1643, in load
    mtime = os.stat(path).st_mtime
FileNotFoundError: [Errno 2] No such file or directory: '/Users/rkn/anaconda3/lib/python3.6/site-packages/s3transfer-0.1.10-py3.6.egg'
```

Uninstalling it with `pip uninstall boto3` and reinstalling it worked.

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**